### PR TITLE
feat: add codegen types support

### PR DESCRIPTION
**Pull request by MateWW**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [CodegenTypes.js](https://github.com/facebook/react-native/blob/main/Libraries/Types/CodegenTypes.js)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/react-native/Codegen.d.ts [hunk 1]
```diff
@@ -29,3 +27,33 @@ declare module 'react-native/Libraries/Utilities/codegenNativeComponent' {
...
+    export type UnsafeObject = object;
+
+    type DefaultTypes = number | boolean | string | ReadonlyArray<string>;
+    // Default handling, ignore the unused value
+    // we're only using it for type checking
+    //
+    // TODO: (rickhanlonii) T44881457 If a default is provided, it should always be optional
+    //  but that is currently not supported in the codegen since we require a default
+    //
+    // eslint-disable-next-line no-unused-vars
```
Comment (eps1lon) [inline types/react-native/Codegen.d.ts:57]: Is this used here?

Comment (MateWW) [reply types/react-native/Codegen.d.ts:57]: Good catch, removed thanks

**Review by eps1lon — APPROVED** [review]

**Comment by kelset** [comment]
cc @ZihanChen-MSFT & @lunaleaps can you check this one out?

**Review by gabritto — APPROVED** [review]

File: types/react-native/Codegen.d.ts [hunk 2]
```diff
@@ -29,3 +27,32 @@ declare module 'react-native/Libraries/Utilities/codegenNativeComponent' {
...
+    export type UnsafeObject = object;
+
+    type DefaultTypes = number | boolean | string | ReadonlyArray<string>;
+    // Default handling, ignore the unused value
+    // we're only using it for type checking
+    //
+    // TODO: (rickhanlonii) T44881457 If a default is provided, it should always be optional
+    //  but that is currently not supported in the codegen since we require a default
+
+    export type WithDefault<Type extends DefaultTypes, Value extends Type | string | undefined> = Type | undefined;
```
Comment (ZihanChen-MSFT) [inline types/react-native/Codegen.d.ts:57]: Flow's optional type `?T` means `T | undefined | null`, it would be better to have `null` also listed here.

**Review by ZihanChen-MSFT — CHANGES_REQUESTED** [review]

File: types/react-native/Codegen.d.ts (hunk 2 above)
Comment (MateWW) [reply types/react-native/Codegen.d.ts]: Good catch, fixed thanks

**Review by ZihanChen-MSFT — APPROVED** [review]

**Comment by MateWW** [comment]
Ready to merge

**Pull request merged**